### DISCUSSION TITLE
How to Shutdown App

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#10200] [Q&A]
What is the correct action to take on a mobile app when the user presses the Back or Close button in the first View?

IClassicDesktopStyleApplicationLifetime has Shutdown method; but ISingleViewApplicationLifetime doesn't. Does setting MainView to Null do the trick?

```c#
if (Application.Current?.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktopApp)
{
    desktopApp.Shutdown();
}
else if (Application.Current?.ApplicationLifetime is ISingleViewApplicationLifetime viewApp)
{
    viewApp.MainView = null;
}
```

**Comment by maxkatz6** [comment]
Not sure what's the best way to close the app on mobile.
Seems like there are different APIs, some of which platform specific, https://stackoverflow.com/questions/29257929/how-to-terminate-a-xamarin-application

**Reply by mysteryx93** [reply]
That's one of those simple questions that we're like... oh... wait a minute... not so simple...

That can allow to close the app.

> However, unlike other iOS methods like calling exit(0) or calling private iOS methods like "terminateWithSuccess" via a selector, it shouldn't fail app store validation purely based on how you do it. They may still fail you because your app tries to terminate itself.
> 
> Either way, you should really question why you need to provide this option. Unlike desktop applications where closing an application is needed because apps reside inside windows which by design allow multi-tasking and are not task orientated, mobile platforms are primarily designed for users to focus on one task at a time. Once the user is finished the task, they should decide to exit this task by clicking the home button, back button or change app (task) button. This really applies to all platforms.

In the context of MvvmDialogs, when you press Back in a dialog, you go back to the main page, but what action should we take when pressing Back from the main page? Nothing and let the OS handle it?

In Xamarin, we override and return True/False to allow/block back navigation. I guess I'll just let the OS handle it without closing the app when there's nothing to move back to. Correct? Except that here we'll need to handle button clicks via API for Android/iOS
```c#
protected override bool OnBackButtonPressed()
{
    if (Condition)
        return true;
    return base.OnBackButtonPressed();
}
```

**Reply by JCH2k** [reply]
Just solved this in my multiplatform Avalonia App:

```
    public override void OnFrameworkInitializationCompleted()
    {
        if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
        {
            ...
        }
        else if (ApplicationLifetime is ISingleViewApplicationLifetime singleViewPlatform)
        {
            var mainView = new MainView
            {
                DataContext = ...
            };
            singleViewPlatform.MainView = mainView;

            RegisterBackNavigation(mainView);
        }

        base.OnFrameworkInitializationCompleted();
    }

    private void RegisterBackNavigation(MainView mainView)
    {
        var topLevel = TopLevel.GetTopLevel(mainView);
        topLevel?.BackRequested += (_, args) =>
        {
            if (MainViewModel?.NavigateBackCommand is { } backCommand && backCommand.CanExecute(null))
            {
                backCommand.Execute(null);
                args.Handled = true;
            }
        };
    }
```

My navigation works in a way that the MainViewModel is handling it. It has a `NavigateBackCommand` where  `CanExecute` returns false if user is on home screen.
So normally this command is handling the back button (`Handled = true`), but if the MainViewModel cann not navigate back, it is not handled and the App is closing.
